src/core/providers/ogr/qgsogrproviderutils.cpp

```diff
@@ -1300,6 +1300,32 @@ QString QgsOgrProviderUtils::quotedValue( const QVariant &value )
 
 OGRLayerH QgsOgrProviderUtils::setSubsetString( OGRLayerH layer, GDALDatasetH ds, QTextCodec *encoding, const QString &subsetString )
 {
+  // Remove any comments
+  QStringList lines {subsetString.split( QChar( '\n' ) )};
+  lines.erase( std::remove_if( lines.begin(), lines.end(), []( const QString & line )
+  {
+    return line.startsWith( QStringLiteral( "--" ) );
+  } ), lines.end() );
+  for ( auto &line : lines )
+  {
+    bool inLiteral {false};
+    QChar literalChar { ' ' };
+    for ( int i = 0; i < line.length(); ++i )
+    {
+      if ( ( ( line[i] == QChar( '\'' ) || line[i] == QChar( '"' ) ) && ( i == 0 || line[i - 1] != QChar( '\\' ) ) ) && ( line[i] != literalChar ) )
```

Comment:
Hate to be annoying, but won't this still be broken for some cases? e.g.

    select "my--thing's" from something where "thing" = 'my "things -- all'

(Of course, anyone who has a database like that needs a good punch in the face)

Reply:
Actually, that case was already handled correctly, I added it to the test: https://github.com/qgis/QGIS/pull/45797/commits/68d69cb7d1a243276213a58c8f625d8bf3680159#diff-56354e2446fe2cb6d1ee92d4e984091172e964e90f3be4d3d42276e033c4986eR2370